### PR TITLE
Plasma Cash UX Improvements - Offline Signing

### PR DESCRIPTION
**Pull request by gakonst**
Implements:
- **Offline signing of transactions**: Removes the limitation of having our keys on the connecting node. This is a necessary step for users wanting to connect and interact with the deployed contracts via infura. 
- **Caching middleware**: Fetch data regarding the coins owned by a user from a local json "database" if they have them already, otherwise poll the node and save locally.
- Minor bug fixes regarding BNs & formatting
- Stateful client API

### CODE REVIEW

File: src/solidity-helpers.ts [hunk 1]
```diff
@@ -12,15 +13,15 @@ export function soliditySha3(...values: any[]): string {
  */
 export class Web3Signer {
```
Comment (enlight) [inline src/solidity-helpers.ts]: You should create a new `OfflineWeb3Signer` class and leave the current `Web3Signer` unchanged.

File: package.json [hunk 2]
```diff
@@ -28,12 +28,15 @@
     "prepublish": "yarn proto && yarn plasmaabi && yarn build && webpack"
   },
   "dependencies": {
+    "@types/lowdb": "^1.0.5",
```
Comment (enlight) [inline package.json]: This should be in devDependencies

File: src/plasma-cash/entity.ts [hunk 3]
```diff
@@ -65,6 +71,35 @@ export class Entity {
     this._childBlockInterval = params.childBlockInterval
   }
 
+  // This should be called whenever a new block gets received
+  // if there is no database we should not allow this to be called
+  async refresh() {
```
Comment (enlight) [inline src/plasma-cash/entity.ts]: refreshAsync

File: package.json [hunk 4]
```diff
@@ -28,12 +28,15 @@
...
     "axios": "^0.18.0",
     "bn.js": "^4.11.8",
     "crypto-browserify": "^3.12.0",
     "debug": "^3.1.0",
     "ethereumjs-util": "^5.2.0",
     "google-protobuf": "3.6",
+    "lodash": "^4.17.11",
```
Comment (enlight) [inline package.json]: lol, how complicated are the two functions you import from lodash? Is it worth pulling in the entirety of lodash?

Comment (gakonst) [reply package.json]: Good point, wasn't even aware of how big lodash is in the first place

**Pull request merged**